opencv/opencv
# WIP: [GSoC] GPU-enabled OpenCV.js, add webgpu-dawn support for dnn module.

### Overview

- Proposal : [GPU-enabled OpenCV.js](https://summerofcode.withgoogle.com/dashboard/project/5352882720735232/details/)

- Mentor : Zhiwen Wu @wzw-intel 

- Student : Hao Jiang [@NALLEIN](https://github.com/NALLEIN)
### This pull request changes

#### **Done**

- Use [Dawn , a  webgpu implemention ](https://dawn.googlesource.com/dawn)to implement some operators under the opencv/modiles/dnn module, refer to [the implementation of Vulkan](https://github.com/opencv/opencv/pull/12703) and reuse the shader. Passed the corresponding test cases on ubuntu.

```
git clone https://github.com/opencv/opencv_extra.git
export OPENCV_TEST_DATA_PATH=${PATH_TO}/opencv_extra/testdata

sudo apt-get install libvulkan1 mesa-vulkan-drivers vulkan-utils
git clone https://github.com/NALLEIN/opencv.git
cd opencv/
git checkout -b layerTest origin/layerTest
mkdir build
cd build
cmake -D CMAKE_BUILD_TYPE=Release -DWITH_WEBGPU=ON -D CMAKE_INSTALL_PREFIX=/usr/local ..
make -j8
$(PATH_TO_OPENCV)/build/bin/opencv_test_dnn --gtest_filter=*layers*
```

- Use [emscripten](https://emscripten.org/) to compile Dawn's API into a wasm format that can be executed in the chrome browser. This module can be compiled to JavaScript using `--build_webgpu` flag, E.g ``` python ./platforms/js/build_js.py build_js --build_webgpu```.
Use the latest [emsdk](https://emscripten.org/docs/getting_started/downloads.html)  version 1.40.1, you can refer to [opencv.js tutorials ](https://docs.opencv.org/3.4/d4/da1/tutorial_js_setup.html)to set up the environment.

####**To do** 

- Verify wheher the DNN_BACKEND_WGPU compiled by emscripten can run correctly in chrome.

- Write the unit test and perf test for DNN_BACKEND_WGPU.

### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [ ] I agree to contribute to the project under OpenCV (BSD) License.
- [ ] To the best of my knowledge, the proposed patch is not based on a code under GPL or other license that is incompatible with OpenCV
- [ ] The PR is proposed to proper branch
- [ ] There is reference to original bug report and related work
- [ ] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [ ] The feature is well documented and sample code can be built with the project CMake
